File: main/src/main/java/lavaplayer/manager/AudioPlayerLifecycleManager.java
package lavaplayer.manager;

import lavaplayer.manager.event.AudioEvent;
import lavaplayer.manager.event.AudioEventListener;
import lavaplayer.manager.event.TrackEndEvent;
import lavaplayer.manager.event.TrackStartEvent;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Triggers cleanup checks on all active audio players at a fixed interval.
 */
public class AudioPlayerLifecycleManager implements Runnable, AudioEventListener {
    private static final long CHECK_INTERVAL = 10000;

    private final ConcurrentMap<AudioPlayer, AudioPlayer> activePlayers;
    private final ScheduledExecutorService scheduler;
    private final AtomicLong cleanupThreshold;
    private final AtomicReference<ScheduledFuture<?>> scheduledTask;

    /**
     * @param scheduler        Scheduler to use for the cleanup check task
     * @param cleanupThreshold Threshold for player cleanup
     */
    public AudioPlayerLifecycleManager(ScheduledExecutorService scheduler, AtomicLong cleanupThreshold) {
        this.activePlayers = new ConcurrentHashMap<>();
        this.scheduler = scheduler;
        this.cleanupThreshold = cleanupThreshold;
        this.scheduledTask = new AtomicReference<>();
    }

    /**
     * Initialise the scheduled task.
     */
    public void initialise() {
        ScheduledFuture<?> task = scheduler.scheduleAtFixedRate(this, CHECK_INTERVAL, CHECK_INTERVAL, TimeUnit.MILLISECONDS);
        if (!scheduledTask.compareAndSet(null, task)) {
            task.cancel(false);
        }
    }

    /**
     * Stop the scheduled task.
     */
    public void shutdown() {
        ScheduledFuture<?> task = scheduledTask.getAndSet(null);
        if (task != null) {
            task.cancel(false);
        }
    }

    @Override
    public void onEvent(@NotNull AudioEvent event) {
        if (event instanceof TrackStartEvent startEvent) {
            activePlayers.put(startEvent.getPlayer(), startEvent.getPlayer());
        } else if (event instanceof TrackEndEvent endEvent) {
            activePlayers.remove(endEvent.getPlayer());
        }
    }

    @Override
    public void run() {
        for (AudioPlayer player : activePlayers.keySet()) {
            player.checkCleanup(cleanupThreshold.get());
        }
    }
}

File: lavaplayer/src/main/java/com/sedmelluq/discord/lavaplayer/manager/AudioPlayerLifecycleManager.kt
package com.sedmelluq.discord.lavaplayer.manager

import com.sedmelluq.discord.lavaplayer.manager.event.AudioEvent
import com.sedmelluq.discord.lavaplayer.manager.event.AudioEventListener
import com.sedmelluq.discord.lavaplayer.manager.event.TrackEndEvent
import com.sedmelluq.discord.lavaplayer.manager.event.TrackStartEvent
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.InternalCoroutinesApi
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * Triggers cleanup checks on all active audio players at a fixed interval.
 *
 * @param scheduler        Scheduler to use for the cleanup check task
 * @param cleanupThreshold Threshold for player cleanup
 */
@OptIn(InternalCoroutinesApi::class)
class AudioPlayerLifecycleManager(
    private val scheduler: ScheduledExecutorService,
    private val cleanupThreshold: () -> Long
) : Runnable, AudioEventListener {
    companion object {
        private const val CHECK_INTERVAL: Long = 10000
    }

    private val activePlayers = ConcurrentHashMap<AudioPlayer, AudioPlayer>()
    private var scheduledTask by atomic<ScheduledFuture<*>?>(null)

    /**
     * initialize the scheduled task.
     */
    init {
        val task = scheduler.scheduleAtFixedRate(this, CHECK_INTERVAL, CHECK_INTERVAL, TimeUnit.MILLISECONDS)
        if (scheduledTask == null) {
            scheduledTask = task
            task.cancel(false)
        }
    }

    /**
     * Stop the scheduled task.
     */
    fun shutdown() {
        scheduledTask?.cancel(false)
        scheduledTask = null
    }

    override suspend fun onEvent(event: AudioEvent) {
        when (event) {
            is TrackStartEvent -> activePlayers[event.player] = event.player
            is TrackEndEvent -> activePlayers.remove(event.player)
        }
    }

    override fun run() {
        for (player in activePlayers.keys) {
            player.checkCleanup(cleanupThreshold())
        }
    }
}
